models/public/googlenet-v4-tf/freeze.py.patch

```diff
@@ -0,0 +1,68 @@
+diff --git a/research/slim/freeze.py b/research/slim/freeze.py
+new file mode 100644
+index 0000000..411001d
+--- /dev/null
++++ b/research/slim/freeze.py
+@@ -0,0 +1,62 @@
++import argparse
++import os
++import tensorflow as tf
++from tensorflow.python.framework import graph_io
++import subprocess
++
++from nets import nets_factory
++
++def parse_args():
++    """Parse input arguments"""
++
++    parser = argparse.ArgumentParser(description='Freeze models from TF Slim')
++
++    parser.add_argument('--ckpt', type=str, required=True,
++                        help='Path to .ckpt file')
++    parser.add_argument('--name', type=str, required=True,
++                        help='Model name')
++    parser.add_argument('--num_classes', type=int, required=True,
++                        help='Number of classes')
++    parser.add_argument('--output', type=str, required=False,
++                        help='Name of output node')
++    return parser.parse_args()
++
++def convert(model_name, ckpt, num_classes, output=None, size=None):
++    tf.logging.set_verbosity(tf.logging.INFO)
++
++    with tf.Graph().as_default() as graph:
++        network_fn = nets_factory.get_network_fn(model_name, num_classes=num_classes, is_training=False)
++
++        size = size or network_fn.default_image_size
++        placeholder = tf.placeholder(name='input', dtype=tf.float32, shape=(1, size, size, 3))
++        logits, _ = network_fn(placeholder)
++
++        pb = ckpt.replace('ckpt', 'pb')
++        frozen = ckpt.replace('.ckpt', '_frozen.pb')
++
++        print('Writing {}'.format(pb))
++        graph_io.write_graph(graph, os.path.dirname(pb), pb, as_text=True)
++
++        print('Running `freeze_graph.py`...')
++        logits = logits if isinstance(logits, tuple) else (logits,)
++        output_node = output or ','.join(l.name[:-2] for l in logits)
++        params = [
++            'freeze_graph',
++            '--input_graph={}'.format(pb),
++            '--input_checkpoint={}'.format(ckpt),
++            '--input_binary=False',
++            '--output_graph={}'.format(frozen),
++            '--output_node_names={}'.format(output_node),
++        ]
++        print(' '.join(params))
++        subprocess.check_call(params)
```

Comment:
Couldn't you just call `tf.graph_util.convert_variables_to_constants` directly and then save the result? Should be simpler than this.

Reply:
To call `convert_variables_to_constants` we first need to restore variables from `.ckpt` file, in my view using `freeze_graph` is easier than implementing its logic.

Reply:
Is it really easier? I just tried replacing the `freeze_graph` invocation with `convert_variable_to_constants`, and this is what I got:

```python
    with tf.Graph().as_default() as graph, tf.Session() as sess:
        network_fn = nets_factory.get_network_fn(model_name, num_classes=num_classes, is_training=False)

        size = size or network_fn.default_image_size
        placeholder = tf.placeholder(name='input', dtype=tf.float32, shape=(1, size, size, 3))
        logits, _ = network_fn(placeholder)

        frozen = ckpt.replace('ckpt', 'frozen.pb')

        logits = logits if isinstance(logits, tuple) else (logits,)
        output_node = output or ','.join(l.name[:-2] for l in logits)

        tf.train.Saver().restore(sess, ckpt)
        frozen_graph_def = tf.graph_util.convert_variables_to_constants(
            sess, graph.as_graph_def(), [output_node])

        print('Writing {}'.format(frozen))
        graph_io.write_graph(frozen_graph_def, os.path.dirname(frozen), frozen, as_text=False)
```

This is shorter than the original code, and better in a few other ways:

1. It doesn't rely on `freeze_graph` being in `PATH`.
2. It doesn't produce an extra output file.